### PR TITLE
Slider

### PR DESCRIPTION
**Pull request by YonathanMeguira**
Added labels on slider:
Labels datoTooltip - like should appear when dragging cursors
Regular labels should replace them when not dragging

No use of datoTooltip as they don't move along while dragging

Added tests

### CODE REVIEW

File: lib/src/slider/slider.spec.ts [hunk 1]
```diff
@@ -1,37 +1,173 @@
...
+import { Component } from '@angular/core';
+
+@Component({ selector: 'custom-host', template: '' })
+class CustomHostComponent {
+  single = new FormControl(10);
+  wrongForm = new FormControl(-300);
+  multi = new FormControl({ min: 10, max: 30 });
+}
+
+fdescribe('slider', () => {
```
Comment (NetanelBasal) [inline lib/src/slider/slider.spec.ts]: remove fdescribe and please write tests for all possibles cases, we need something like 80% coverage at least.

File: lib/src/slider/slider.component.ts [hunk 2]
```diff
@@ -189,24 +218,82 @@ export class DatoSliderComponent extends BaseCustomControl implements AfterViewI
...
+    const maxCursorLeave$ = fromEvent(this.maxCursorElement, 'mouseleave');
+
+    const eventFilter = () => this.showLabels && !this.minCursorMoving && !this.maxCursorMoving;
+
+    minCursorLeave$
+      .pipe(
+        filter(eventFilter),
+        untilDestroyed(this)
+      )
+      .subscribe(() => this.minLabelElement.classList.remove('label-moving'));
```
[the remark this answers is not in the document]
Comment (YonathanMeguira) [reply lib/src/slider/slider.component.ts]: done

File: lib/src/slider/slider.spec.ts [hunk 3]
```diff
@@ -1,37 +1,173 @@
...
+  });
+
+  const options = {
+    min: 0,
+    max: 100,
+    step: 1,
+    multiCursor: false,
+    disabled: false,
+    unit: ''
+  };
```
Comment (NetanelBasal) [inline lib/src/slider/slider.spec.ts]: each it needs to start with the word `should`

File: lib/src/slider/slider.spec.ts [hunk 4]
```diff
@@ -1,37 +1,173 @@
...
+
+  it('value passed to the form should be reflected on the slider', function() {
+    host = createHost(`<div style="width: 300px"><dato-slider [formControl]="single"></dato-slider></div>`, false);
+    host.component.options = options;
+    host.detectChanges();
+    const unitSize = host.component['unitSize'];
+    const sliderStart = host.query('.slider-wrapper').getBoundingClientRect().left;
```
Comment (NetanelBasal) [inline lib/src/slider/slider.spec.ts]: put every selector string in a const and use it.

File: lib/src/slider/slider.spec.ts [hunk 5]
```diff
@@ -1,37 +1,173 @@
...
+    host.dispatchMouseEvent(minCursor, 'mouseup');
+
+    host.dispatchMouseEvent(maxCursor, 'mousedown');
+    host.dispatchMouseEvent(maxCursor, 'mousemove', maxProgression);
+    host.dispatchMouseEvent(maxCursor, 'mouseup');
+
+    const minCursorX = minCursor.getBoundingClientRect().left;
+    const maxCursorX = maxCursor.getBoundingClientRect().left;
+
+    console.log(minCursorX, maxCursorX);
```
Comment (NetanelBasal) [inline lib/src/slider/slider.spec.ts]: remove

File: lib/src/slider/slider.spec.ts [hunk 6]
```diff
@@ -1,37 +1,173 @@
...
+
+    host.detectChanges();
+
+    const sliderStart = host.query('.slider-wrapper').getBoundingClientRect().left;
+    const minProgression = sliderStart + 70;
+    const maxProgression = sliderStart + 20;
+
+    const minCursor = host.query('.cursor-container:first-child');
+    const maxCursor = host.query('.cursor-container:last-child');
+
```
Comment (NetanelBasal) [inline lib/src/slider/slider.spec.ts]: great tests but there is a lot more to check, also check the form control value updates.

File: lib/src/slider/slider.component.ts [hunk 7]
```diff
@@ -463,6 +557,32 @@ export class DatoSliderComponent extends BaseCustomControl implements AfterViewI
     this.sliderDimensions = this.slider.nativeElement.getBoundingClientRect();
   }
 
+  /**
+   * updates the label value on each mousemove || afterViewInit
+   */
+  private updateLabels() {
+    if (this.showLabels) {
+      if (this.minCursorMoving) {
+        this.minLabelElement.innerText = `${this.cursorsState.min.model}${this.options.unit}`;
```
[the remark this answers is not in the document]
Comment (YonathanMeguira) [reply lib/src/slider/slider.component.ts]: done

File: lib/src/slider/slider.component.ts [hunk 8]
```diff
@@ -189,24 +218,82 @@ export class DatoSliderComponent extends BaseCustomControl implements AfterViewI
...
+        filter(eventFilter),
+        untilDestroyed(this)
+      )
+      .subscribe(() => this.minLabelElement.classList.remove('label-moving'));
+    maxCursorLeave$
+      .pipe(
+        filter(eventFilter),
+        untilDestroyed(this)
+      )
+      .subscribe(() => this.maxLabelElement.classList.remove('label-moving'));
```
[the remark this answers is not in the document]
Comment (YonathanMeguira) [reply lib/src/slider/slider.component.ts]: ok

File: lib/src/slider/slider.component.ts [hunk 9]
```diff
@@ -189,24 +219,82 @@ export class DatoSliderComponent extends BaseCustomControl implements AfterViewI
...
+        this.minLabelElement.innerText = `${this.cursorsState.min.model}${this.options.unit}`;
+        this.minLabelElement.classList.add(this.movingLabelClass);
+      });
+    maxHover$
+      .pipe(
+        filter(filterEvent),
+        untilDestroyed(this)
+      )
+      .subscribe(() => {
+        this.maxLabelElement.innerText = `${this.cursorsState.max.model}${this.options.unit}`;
```
Comment (bh86) [inline lib/src/slider/slider.component.ts]: use `setMaxLabel` function.

File: lib/src/slider/slider.component.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bh86) [reply lib/src/slider/slider.component.ts]: you forgot the min function :)

File: lib/src/slider/slider.component.ts (hunk 9 above)
Comment (YonathanMeguira) [reply lib/src/slider/slider.component.ts]: done

**Pull request merged**